src/core/Oscillator.cpp

```diff
@@ -88,6 +106,238 @@ void Oscillator::update( sampleFrame * _ab, const fpp_t _frames,
+	}
+}
+
+
+
+sample_t Oscillator::s_waveTables[Oscillator::WaveShapes::NumWaveShapes-2][128 / OscillatorConstants::SEMITONES_PER_TABLE][OscillatorConstants::WAVETABLE_LENGTH];
+fftwf_plan Oscillator::s_fftPlan;
+fftwf_plan Oscillator::s_ifftPlan;
+fftwf_complex * Oscillator::s_specBuf;
+float Oscillator::s_fftBuffer[OscillatorConstants::WAVETABLE_LENGTH*2];
+float Oscillator::s_sampleBuffer[OscillatorConstants::WAVETABLE_LENGTH];
+
+
+
+void Oscillator::createFFTPlans()
+{
+	Oscillator::s_specBuf = ( fftwf_complex * ) fftwf_malloc( ( OscillatorConstants::WAVETABLE_LENGTH * 2 + 1 ) * sizeof( fftwf_complex ) );
+	Oscillator::s_fftPlan = fftwf_plan_dft_r2c_1d(OscillatorConstants::WAVETABLE_LENGTH, s_sampleBuffer, s_specBuf, FFTW_MEASURE );
+	Oscillator::s_ifftPlan = fftwf_plan_dft_c2r_1d(OscillatorConstants::WAVETABLE_LENGTH, s_specBuf, s_sampleBuffer, FFTW_MEASURE);
+	// initialize s_specBuf content to zero, since the values are used in a condition inside generateFromFFT()
+	for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH * 2 + 1; i++)
+	{
+		s_specBuf[i][0] = 0.0f;
+		s_specBuf[i][1] = 0.0f;
+	}
+}
+
+void Oscillator::destroyFFTPlans()
+{
+	fftwf_destroy_plan(s_fftPlan);
+	fftwf_destroy_plan(s_ifftPlan);
+	fftwf_free(s_specBuf);
+}
+
+void Oscillator::generateWaveTables()
+{
+	//generate sine tables
+	auto sinGen = []()
+	{
+		for (int i = 0; i < OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT; ++i)
+		{
+			generateSineWaveTable(s_waveTables[WaveShapes::SineWave][i]);
+		}
+	};
+
+	//generate saw tables
+	auto sawGen = []()
+	{
+		for (int i = 0; i < OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT; ++i)
+		{
+			generateSawWaveTable(OscillatorConstants::MAX_FREQ / freqFromWaveTableBand(i), s_waveTables[WaveShapes::SawWave][i]);
+		}
+	};
+	//generate square tables
+	auto squareGen = []()
+	{
+		for (int i = 0; i < OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT; ++i)
+		{
+			generateSquareWaveTable(OscillatorConstants::MAX_FREQ / freqFromWaveTableBand(i), s_waveTables[WaveShapes::SquareWave][i]);
+		}
+	};
+	//generate triangle tables
+	auto triangleGen = []()
+	{
+		for (int i = 0; i < OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT; ++i)
+		{
+			generateTriangleWaveTable(OscillatorConstants::MAX_FREQ / freqFromWaveTableBand(i), s_waveTables[WaveShapes::TriangleWave][i]);
+		}
+	};
+	auto fftGen = []()
+	{
+		//generate moogSaw tables
+		//generate signal buffer
+		for (int i = 0; i < OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT; ++i)
+		{
+			for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH; ++i)
+			{
+				Oscillator::s_sampleBuffer[i] = moogSawSample((float)i / (float)OscillatorConstants::WAVETABLE_LENGTH);
+			}
+			fftwf_execute(s_fftPlan);
+			generateFromFFT(OscillatorConstants::MAX_FREQ / freqFromWaveTableBand(i), 0.2f, s_waveTables[WaveShapes::MoogSawWave][i]);
```

Comment:
The harmonics for square wave drop off very slowly with higher frequency, so they essentially never reach even the higher threshold. OTOH, Exponential wave drops off quickly and even 0.001 is too high to prevent cutting off some high harmonics in some cases.